### PR TITLE
GG-30804

### PR DESCRIPTION
**Pull request by sergeyuttsel**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 1]
```diff
@@ -619,6 +619,8 @@ private void onDiscovery0(DiscoveryNotification notification) {
                     boolean incMinorTopVer;
 
                     if (customMsg instanceof ChangeGlobalStateMessage) {
+                        ctx.state().onStateMessage();
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:622]: We don't need this new method in ClusterStateProcessor: it's okay to add this piece of logic right into existing onStateChangeMessage.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 2]
```diff
@@ -626,13 +626,17 @@ private void onDiscoveryEvent(DiscoveryEvent evt, DiscoCache cache) {
                     else {
                         DiscoveryDataClusterState state = cctx.kernalContext().state().clusterState();
 
+                        assert state.transition() : state + " evt: " + evt + " msg: " + customMsg;
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java:629]: Why do we need this assertion? What does it protect us from?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java [hunk 3]
```diff
@@ -176,6 +179,12 @@
     /** */
     private final ConcurrentMap<UUID, GridFutureAdapter<Void>> transitionFuts = new ConcurrentHashMap<>();
 
+    /**
+     * This latch is used to ensure that on client nodes the {@link ChangeGlobalStateMessage} is fully processed
+     * in the {@link GridCachePartitionExchangeManager} before the {@link ChangeGlobalStateFinishMessage} is processed.
+     */
+    private volatile CountDownLatch changeStateInProgress;
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java:186]: I reviewed how StateChange messages are created and sent, it seems possible to get two consequent StateChange messages before receiving StateChangeFinished for the first one.

So single latch isn't enough here, we should use ConcurrentHashMap and map latches by request IDs available both in StateChange and StateChangeFinished messages.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java [hunk 4]
```diff
@@ -622,6 +631,24 @@ private boolean isBaselineSatisfied(BaselineTopology blt, List<ClusterNode> srvN
     @Override public void onStateFinishMessage(ChangeGlobalStateFinishMessage msg) {
         DiscoveryDataClusterState discoClusterState = globalState;
 
+        if (ctx.clientNode() && !ctx.isDaemon() && changeStateInProgress != null) {
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java:634]: I don't like this style when we inline a very specific logic as plain blocks of code into existing methods.

Why not to create a new class called for instance ClientStateChangeSynchronizer and encapsulate this logic there (including aforementioned ConcurrentHashMap)?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java [hunk 5]
```diff
@@ -662,6 +689,23 @@ private boolean isBaselineSatisfied(BaselineTopology blt, List<ClusterNode> srvN
...
     }
 
+    /** */
+    public void onStateMessage() {
+        if (ctx.clientNode() && !ctx.isDaemon())
+            changeStateInProgress = new CountDownLatch(1);
+    }
+
+    /** */
+    public void onChangeStateMessageWasProcessed() {
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java:699]: I suggest to rename this method to `onStateMessageProcessingFinished`.

**Review by sergey-chugunov-1985 — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 2 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java:629]: This line was removed in https://github.com/gridgain/gridgain/commit/c932f681e53c1b5d5fe5c637d5e387ded8c6443c#diff-896382ae47330cb2c3aa1505290c99e18c93c338b9e584106d2e9e14e00a4b92 and a ticket GG-30804 was created. So now I revert it.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java (hunk 1 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:622]: `onStateChangeMessage` invokes from other code. But we need to invoke `onStateMessage()` only for this case.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java (hunk 5 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java:699]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java (hunk 4 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java:634]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java (hunk 3 above)
Comment (sergeyuttsel) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java:186]: Done

**Pull request closed without merging**